mlir/lib/Catalyst/Utils/CountStaticForloopIterations.cpp

```diff
@@ -0,0 +1,83 @@
+// Copyright 2025 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cmath> // std::ceil()
+
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Operation.h"
+
+using namespace mlir;
+
+namespace catalyst {
+
+static int64_t getNumIterations(int64_t lowerBound, int64_t upperBound, int64_t step)
+{
+    return std::ceil((upperBound - lowerBound) / step);
```

Comment:
wouldn't we lose precision here before the result is passed to `std::ceil`? e.g. ceil(5/2) would still be calculated to 2 since they are integers.
```suggestion
    return std::ceil((double)(upperBound - lowerBound) / step);
```
or alternatively
```suggestion
    return (upperBound - lowerBound + step - 1) / step;
```

Reply:
Also we should probably verify that `step` is not zero and `upperBound` is larger than `lowerBound`. This is definitely handled by scf dialect but it still good to have since this is programmed as a util function.

Reply:
well we first evaluate 5 / 2 with is 2 if 5 and 2 are integers and then passthe result to ceil and ceil(2) is 2. Am I missing something? 🤔

Reply:
> well we first evaluate 5 / 2 with is 2 if 5 and 2 are integers and then passthe result to ceil and ceil(2) is 2. Am I missing something? 🤔

I will add this (learned something new about cpp today)

Reply:
> Also we should probably verify that `step` is not zero and `upperBound` is larger than `lowerBound`. This is definitely handled by scf dialect but it still good to have since this is programmed as a util function.

https://mlir.llvm.org/docs/Dialects/SCFDialect/#scffor-scfforop

> The step is a value of same [index] type but required to be positive, the lower and upper bounds can be also negative or zero. 

Since step is guaranteed to be positive I'd say these conditions are automatic. But yes I will assert them here regardless.

Reply:
https://github.com/PennyLaneAI/catalyst/pull/1863/commits/ac4253ef05313fcbcca74ecdfd7769b60dc04503